postreise/plot/plot_bar_shortfall.py

```diff
@@ -0,0 +1,223 @@
+
+    :param list/str areas: list of area(s) to show shortfall bar plots. If the target of
+        an area is not defined in target_df, it will be ignored.
+    :param int/list/str scenario_ids: list of scenario id(s).
+    :param pandas.DataFrame target_df: target data frame, which defines the clean
+        energy target fraction, allowed resources and external historical amount of
+        qualified energy for each area.
+    :param dict strategy: a dictionary with keys being scenario ids and values being
+        strategies, either *"collaborative"* or *"independent"*.
+    :param list/str scenario_names: list of scenario name(s) of same len as scenario
+        ids, defaults to None.
+    :param str/int baseline_scenario: scenario id that serves as a baseline in the
+        bar chart, default to None.
+    :param str baseline_scenario_name: specify the label of the baseline scenario
+        shown in the bar chart, default to None, in which case the name of the
+        scenario will be used.
+    :raises ValueError:
+        if length of scenario_names and scenario_ids is different.
+    :raises TypeError:
+        if target_df is not a pandas.DataFrame and/or
+        if strategy is provided but not in a dictionary format and/or
+        if baseline_scenario is provided but not in a str/int format and/or
+        if baseline_scenario_name is provided but not in a str format.
+    """
+    if isinstance(areas, str):
+        areas = [areas]
+    if isinstance(scenario_ids, (int, str)):
+        scenario_ids = [scenario_ids]
+    if not isinstance(target_df, pd.DataFrame):
+        raise TypeError("ERROR: target_df must be a pandas.DataFrame")
+    if not strategy:
+        strategy = dict()
+    if not isinstance(strategy, dict):
+        raise TypeError("ERROR: strategy must be a dictionary")
+    if isinstance(scenario_names, str):
+        scenario_names = [scenario_names]
+    if scenario_names and len(scenario_names) != len(scenario_ids):
+        raise ValueError(
+            "ERROR: if scenario names are provided, number of scenario names must match number of scenario ids"
+        )
+    if baseline_scenario and not isinstance(baseline_scenario, (str, int)):
+        raise TypeError("ERROR: baseline_scenario must be a string or integer")
+    if baseline_scenario_name and not isinstance(baseline_scenario_name, str):
+        raise TypeError("ERROR: baseline_scenario_name must be a string")
+
+    scenarios = dict()
+    targets = dict()
+    all_sids = scenario_ids + [baseline_scenario] if baseline_scenario else scenario_ids
+    for sid in all_sids:
+        s = Scenario(sid)
+        tmp_df = target_df.copy()
+        tmp_df = add_resource_data_to_targets(tmp_df, s)
+        tmp_df = add_demand_to_targets(tmp_df, s)
+        tmp_df = add_shortfall_to_targets(tmp_df)
+        scenarios[sid] = s
+        targets[sid] = tmp_df
+    for area in areas:
+        if area not in target_df.index and area != "all":
+            continue
```

Comment:
Do we want to at least warn the user when one or more areas is being skipped?

Reply:
Done.